### ISSUE TITLE
Issues with Textual Chunks

### ISSUE THREAD

**Issue by leonbloy** [leonbloy/pngcs#5]
_From [taurenje...@gmail.com](https://code.google.com/u/112390715196505478818/) on November 27, 2012 22:53:51_

【problem 1】
PngChunkTEXT.ParseFromRaw(ChunkRaw c) throws a exception, because the length of String[] k is 1,
where k[0] is "Software\0Adobe....".

I modified the code, and it works well:

// Updated by Jerin@2012-11-28
//String[] k = Hjg.Pngcs.PngHelperInternal.charsetLatin1.GetString(c.Data).Split(new char[] { '\0' });
String[] k = Hjg.Pngcs.PngHelperInternal.charsetLatin1.GetString( c.Data ).Split( new string[] { @"\0" }, StringSplitOptions.None );

【problem 2】
My scenario is to add/read a custom metadata of png.

I browsed samples of the project but found no way to read chunks only -- SampleCustomChunk/Program.cs/testRead(String file) gives an advice: call pngr.GetRow(pngr.ImgInfo.Rows - 1) first.

so i add a public method in PngReader class:

public void ReadChunksOnly() {
  if ( FirstChunksNotYetRead() )
    ReadFirstChunks();
  ReadLastChunks();
}

Thanks for your work very much!

_Original issue: http://code.google.com/p/pngcs/issues/detail?id=5_

**Comment by leonbloy** [comment]
_From [taurenje...@gmail.com](https://code.google.com/u/112390715196505478818/) on November 27, 2012 18:08:58_

cannot edit the issue -_-

update the problem 1:

String[] k = Hjg.Pngcs.PngHelperInternal.charsetLatin1.GetString(c.Data).Split(new char[] { '\0' });
// Updated by Jerin@2012-11-28
if( k.Length == 1)
   k = Hjg.Pngcs.PngHelperInternal.charsetLatin1.GetString( c.Data ).Split( new string[] { @"\0" }, StringSplitOptions.None );

**Comment by leonbloy** [comment]
_From [taurenje...@gmail.com](https://code.google.com/u/112390715196505478818/) on November 27, 2012 18:11:49_

It seems like that:
for the original png, using char[] works well
but when i add custom chunk, i must use string[]

here is my sample code:

public const string Key_SemanticInfo = "SemanticInfo";

using ( FileStream newFileStream = new FileStream( saveDialog.FileName, FileMode.OpenOrCreate, FileAccess.ReadWrite ),
          oldFileStream = new FileStream( soureFilePath, FileMode.Open, FileAccess.Read ) ) {
            string semanticInfo = @"
    sky: blue sky.
    heart: cloud heart.
    Meaning: It means love|爱|❤.
                ";

```
      ////// Reference: https://code.google.com/p/pngcs/wiki/Overview // get decoder
      pngReader = new PngReader( oldFileStream );
      // create encoder
      pngWriter = new PngWriter( newFileStream, pngReader.ImgInfo );
      // copy
      int chunkBehav = Hjg.Pngcs.Chunks.ChunkCopyBehaviour.COPY_ALL; // tell to copy all 'safe' chunks
      pngWriter.CopyChunksFirst( pngReader, chunkBehav );          // copy some metadata from reader 
      int channels = pngReader.ImgInfo.Channels;
      if ( channels \< 3 )
        throw new Exception( "This example works only with RGB/RGBA images" );
      for ( int row = 0; row \< pngReader.ImgInfo.Rows; row++ ) {
        ImageLine l1 = pngReader.ReadRow( row ); // format: RGBRGB... or RGBARGBA...
        pngWriter.WriteRow( l1, row );
      }
      pngWriter.CopyChunksLast( pngReader, chunkBehav ); // metadata after the image pixels? can happen

      // app info
      pngWriter.GetMetadata().SetText( Hjg.Pngcs.Chunks.PngChunkTextVar.KEY_Software, "Semantic Image" );
      // semantic info
      Hjg.Pngcs.Chunks.PngChunk chunk = pngWriter.GetMetadata().SetText( Key_SemanticInfo, semanticInfo, false, false );
      chunk.Priority = true;

      pngWriter.End(); // dont forget this
      pngReader.End();

      oldFileStream.Close();
      newFileStream.Close();
    }
```

**Comment by leonbloy** [comment]
_From [hgonzalez@gmail.com](https://code.google.com/u/hgonzalez@gmail.com/) on November 30, 2012 06:17:30_

Thanks for the report, I will give it a look soon.
But the original split seems right to me. The problems seems to be other:  your text is Unicode, hence it should have been stored as PngChunkITXT, not as PngChunkTEXT. It seems that SetText() is to blame, I'll check that.

**Comment by leonbloy** [comment]
_From [hgonzalez@gmail.com](https://code.google.com/u/hgonzalez@gmail.com/) on November 30, 2012 09:38:56_

The fix of comment 1 does not make sense much to me. And I don't know if the code of comment 2 generates a PNG that later gives that exception or what. Could you please post a detailed step by step way of reproducing the problem? You can also attach a problematic PNG. Or you can use the tweakPNG utility to look into the chunks http://entropymine.com/jason/tweakpng/

**Comment by leonbloy** [comment]
_From [hgonzalez@gmail.com](https://code.google.com/u/hgonzalez@gmail.com/) on November 30, 2012 19:56:52_

I have already found the problem, it will be fixed soon. (Just commited in PngJ)

**Comment by leonbloy** [comment]
_From [hgonzalez@gmail.com](https://code.google.com/u/hgonzalez@gmail.com/) on November 30, 2012 19:58:41_

**Status:** Started  
  **Owner:** hgonzalez@gmail.com  
  **Labels:** -Priority-Medium Priority-High

**Comment by leonbloy** [comment]
_From [hgonzalez@gmail.com](https://code.google.com/u/hgonzalez@gmail.com/) on December 02, 2012 11:24:40_

**Summary:** Issues with Textual Chunks  <span class="oldvalue"> (was: 2 problems) </span>

**Comment by leonbloy** [comment]
_From [hgonzalez@gmail.com](https://code.google.com/u/hgonzalez@gmail.com/) on December 08, 2012 14:05:51_

Solved in 1.1.3 . See also the TestTextChunks.

Regarding the scenario in which one wants to read only the metadata (chunks), a new method is provided ReadSkippingAll rows, much efficient.

**Status:** Fixed
